pkg/querier/batch/stream.go

```diff
@@ -159,11 +159,49 @@ func (bs *batchStream) merge(batch *chunk.Batch, size int) {
 			populate(batch, rt)
 			batch.Next()
 		} else {
-			if (rt == chunkenc.ValHistogram || rt == chunkenc.ValFloatHistogram) && lt == chunkenc.ValFloat {
-				// Prefer histograms than floats. Take left side if both have histograms.
-				populate(batch, rt)
-			} else {
+			// The two samples have the same timestamp. They may be the same exact sample or different one
+			// (e.g. we ingested an in-order sample and then OOO sample with same timestamp but different value).
+			// We want query results to be deterministic, so we select one of the two side deterministically.
+			var pickLeftSide bool
+
+			switch {
```

Comment:
i'd break all of your changes into a function. It's an edge case and I think it's distracting from the main body of the function


something like 

```go
func resolveTimestampConflict(lt, rt chunkenc.ValueType, leftBatch, rightBatch *chunk.Batch) (pickedVT chunkenc.ValueType, pickedB, discardedB *chunk.Batch)
```

Reply:
It's not an edge case, it's the most common case as usually two ingesters/store-gateways will have the same data.

Reply:
ah good point. Still think a function would make it more obvious